ouroboros-network/src/Ouroboros/Network/Subscription/PeerState.hs

```diff
@@ -0,0 +1,559 @@
+import           Control.Monad.State
+import           Data.Map.Strict (Map)
+import qualified Data.Map as Map
+import           Data.Maybe (mapMaybe)
+import           Data.List.NonEmpty (NonEmpty (..))
+import           Data.Semigroup (sconcat)
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           Data.Typeable ( Proxy (..)
+                               , (:~:) (..)
+                               , eqT
+                               , gcast
+                               , tyConName
+                               , typeRepTyCon
+                               , typeRep
+                               )
+import           Text.Printf
+
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadTime
+
+import           Data.Semigroup.Action
+
+-- | Semigroup of commands which acts on 'PeerState'.  The @t@ variable might
+-- be initiated to 'DiffTime' or @Time m@.
+--
+-- This semigroup allows to either suspend both consumer and producer or just
+-- the consumer part.
+--
+data SuspendCommand t
+    = SuspendPeer !t !t
+    -- ^ peer is suspend; The first @t@ is the time until which a local
+    -- producer is suspended, the second one is the time until which a local
+    -- consumer is suspended.
+    | SuspendConsumer !t
+    -- ^ suspend local consumer \/ initiator side until @t@ (this mean we are
+    -- not allowing to communicate with the producer \/ responder of a remote
+    -- peer).
+    | Throw
+    -- ^ throw an error from the main thread.
+    deriving (Eq, Ord, Show)
+
+consumerSuspendedUntil :: SuspendCommand t -> Maybe t
+consumerSuspendedUntil (SuspendPeer _ consT)   = Just consT
+consumerSuspendedUntil (SuspendConsumer consT) = Just consT
+consumerSuspendedUntil Throw                   = Nothing
+
+producerSuspendedUntil :: SuspendCommand t -> Maybe t
+producerSuspendedUntil (SuspendPeer prodT _) = Just prodT
+producerSuspendedUntil (SuspendConsumer _) = Nothing
+producerSuspendedUntil Throw               = Nothing
+
+instance Functor SuspendCommand where
+    fmap f (SuspendPeer prodT consT) = SuspendPeer (f prodT) (f consT)
+    fmap f (SuspendConsumer consT) = SuspendConsumer (f consT)
+    fmap _ Throw = Throw
+
+-- | The semigroup instance.  Note that composing 'SuspendPeer' with
+-- 'SuspendConsumer' gives 'SuspendPeer'.  'SuspendPeer' and 'SuspendConsumer'
+-- form a sub-semigroup.
+--
+instance Ord t => Semigroup (SuspendCommand t) where
+    Throw <> _ = Throw
+    _ <> Throw = Throw
+    SuspendPeer prodT consT <> SuspendPeer prodT' consT'
+      = SuspendPeer (prodT `max` prodT') (consT `max` consT')
+    SuspendConsumer consT <> SuspendPeer prodT consT'
+      = SuspendPeer prodT (consT `max` consT')
+    SuspendPeer prodT consT <> SuspendConsumer consT'
+      = SuspendPeer prodT (consT `max` consT')
+    SuspendConsumer consT <> SuspendConsumer consT'
+      = SuspendConsumer (consT `max` consT')
```

Comment:
Note that it does have a unit `SuspendConsumer minBound`, but `SuspendConsumer DiffTime` does not, since there's no `Bounded` instance of `DiffTime`, also this unit does not act as an identity on `PeerState` (so it's not a monoid action).  I think we will not need it, if there's an exception handled by an `ErrorPolicy` some decision must be taken.  If it turns out to be needed we can add a constructor for it and have a monoid action.